uavcan/dsdl/signature.py

```diff
@@ -18,51 +18,84 @@
 # Output xor: 0xFFFFFFFFFFFFFFFF
 # Check: 0x62EC59E3F1A4F00A
 #
+import ctypes
+import logging
+import os
+import sysconfig
 
 
 class Signature:
-    '''
+    """
     This class implements the UAVCAN DSDL signature hash function. Please refer to the specification for details.
-    '''
+    """
     MASK64 = 0xFFFFFFFFFFFFFFFF
     POLY = 0x42F0E1EBA9EA3693
 
+    crc64lib = None
+
+    WARNED_MISSING_DLL = False
+
+    dll_path = os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                            '../fastcrc64.{}.so'.format(sysconfig.get_config_vars('SOABI')[0]))
+    if os.path.exists(dll_path):
+        crc64lib = ctypes.CDLL(dll_path)
+        crc64lib.crc64.restype = ctypes.c_uint64
+        crc64lib.crc64.argtypes = [ctypes.c_char_p, ctypes.c_uint32, ctypes.c_uint64]
+
+        crc64lib.finalize.restype = ctypes.c_uint64
+        crc64lib.finalize.argtypes = [ctypes.c_uint64]
+
     def __init__(self, extend_from=None):
-        '''
+        """
         extend_from    Initial value (optional)
-        '''
+        """
         if extend_from is not None:
             self._crc = (int(extend_from) & Signature.MASK64) ^ Signature.MASK64
         else:
             self._crc = Signature.MASK64
 
     def add(self, data_bytes):
-        '''Feed ASCII string or bytes to the signature function'''
+        """Feed ASCII string or bytes to the signature function"""
+        num_bytes = len(data_bytes)
+
         try:
             if isinstance(data_bytes, basestring):  # Python 2.7 compatibility
                 data_bytes = map(ord, data_bytes)
         except NameError:
             if isinstance(data_bytes, str):  # This branch will be taken on Python 3
-                data_bytes = map(ord, data_bytes)
+                data_bytes = data_bytes.encode()
+
+        if not isinstance(data_bytes, bytes):
+            data_bytes = bytes(data_bytes)
 
-        for b in data_bytes:
-            self._crc ^= (b << 56) & Signature.MASK64
-            for _ in range(8):
-                if self._crc & (1 << 63):
-                    self._crc = ((self._crc << 1) & Signature.MASK64) ^ Signature.POLY
-                else:
-                    self._crc <<= 1
+        if Signature.crc64lib is not None:
+            self._crc = Signature.crc64lib.crc64(data_bytes, num_bytes, self._crc)
+        else:
+            if not Signature.WARNED_MISSING_DLL:
```

Comment:
Why not just warn while loading the DLL?

Reply:
morning brain